kong/plugins/ai-proxy/handler.lua

```diff
@@ -0,0 +1,183 @@
+        end
+
+        local new_response_string, err = ai_driver.from_format(response_body, conf.model, route_type)
+        if err then
+          ngx.status = 500
+          kong.ctx.plugin.parsed_response = fmt(INTERNAL_SERVER_ERROR_TEMPLATE, err)
+        end
+        if new_response_string then
+          -- preserve the same response content type; assume the from_format function
+          -- has returned the body in the appropriate response output format
+          kong.ctx.plugin.parsed_response = new_response_string
+        end
+
+        -- else don't transform it
+      end
+    end
+  end
+
+  kong.log.debug("OUT: ai-proxy/header_filter")
+end
+
+function _M:body_filter(conf)
+  kong.log.debug("IN: ai-proxy/body_filter")
+
+  -- all errors MUST be checked and returned in header_filter
+  -- we should receive a replacement response body from the same thread
+  if kong.ctx.plugin.parsed_response then
+    local is_gzip = kong.response.get_header("Content-Encoding") == "gzip"
+    if is_gzip then
+      kong.ctx.plugin.parsed_response = kong_utils.deflate_gzip(kong.ctx.plugin.parsed_response)
+    end
+    kong.response.set_raw_body(kong.ctx.plugin.parsed_response)
+  end
+
+  -- call with replacement body, or original body if nothing changed
+  ai_shared.post_request(conf, kong.ctx.plugin.parsed_response or kong.response.get_raw_body())
+
+  kong.log.debug("OUT: ai-proxy/body_filter")
+end
+
+function _M:access(conf)
+  kong.log.debug("IN: ai-proxy")
+  kong.service.request.enable_buffering()
+
+  -- store the route_type in ctx for use in response parsing
+  local route_type = conf.route_type
+  kong.ctx.plugin.operation = route_type
+
+  if kong.ctx.shared.prompt_guarded and not kong.ctx.shared.prompt_rendered then
+    do_bad_request("this AI model is prompt-guarded; you must execute a request from the prompt templates")
+  end
+
+  kong.log.debug("using AI request driver ", conf.model.provider)
+  local ai_driver = require("kong.plugins.ai-proxy.drivers." .. conf.model.provider)
+
+  -- first, calculate the coordinates of the request
+  local content_type = kong.request.get_header("Content-Type")
+  if not content_type then
+    -- assume application/json for most requests
+    content_type = "application/json"
+  end
+
+  local request_table, err
+  request_table = kong.request.get_body(content_type)
+  if not request_table then
+    do_bad_request("content-type header does not match request body")
+  end
+  if kong.ctx.shared.prompt_rendered then
+    request_table, err = apply_transformations(route_type, request_table, kong.ctx.shared.transformation_instructions)
+    if err then do_internal_server_error("failed to request transformations: " .. err) end
+  end
+
+  -- TODO: decision: validate directly within this plugin?
+
+  kong.log.inspect("ai-proxy request: ", request_table)
+
+  -- execute pre-request hooks for this driver
+  local err = ai_driver.pre_request(conf, request_table)
+  if err then
+    do_bad_request(err)
```

Comment:
Just another reminder re: `return`-ing when your function will exit the handler function to return a response to the client:

```suggestion
    return bad_request(err)
```

(also dropping the `do_` prefix from the function name makes it read a little nicer to me with the added `return` keyword)

Reply:
Done